### PR TITLE
Only publish masters track on masters seat creation

### PR DESCRIPTION
**Pull request by schenedx**
Update the course discovery so if a Masters seat type is created on course_run, then publish the seat to LMS to create the masters seat in LMS
@edx/masters-neem please review

### CODE REVIEW

File: course_discovery/apps/course_metadata/signals.py [hunk 1]
```diff
@@ -34,17 +39,53 @@ def add_masters_track_on_course(sender, instance, **kwargs):  # pylint: disable=
...
-        if not program.partner.lms_commerce_api_url:
-            return None
-
-        partner = program.partner
-        commerce_api_client = OAuthAPIClient(partner.lms_url, partner.oidc_key, partner.oidc_secret)
+        us_currency = Currency.objects.get(code='USD')
 
         for course_run in instance.course_runs:
-            _add_masters_track_on_commerce_course_run(course_run, partner.lms_commerce_api_url, commerce_api_client)
+            Seat.objects.update_or_create(
```
Comment (reillz10) [inline course_discovery/apps/course_metadata/signals.py:51]: So much cleaner! I love it

File: course_discovery/apps/course_metadata/signals.py [hunk 2]
```diff
@@ -71,18 +112,9 @@ def _add_masters_track_on_commerce_course_run(course_run, api_url, commerce_api_
 
 
 def _seats_for_course_run(course_run):
-    """ Given a course run, this method serializes all current available seats and appends a masters seat. """
-
     seats = course_run.enrollable_seats()
     current_seat_types = [_serialize_seat_as_course_mode_for_commerce_api(seat) for seat in seats]
-    return current_seat_types + [{
```
Comment (reillz10) [inline course_discovery/apps/course_metadata/signals.py:78]: So cool, now our model handles this. Such a nice refactor 🤙

**Review by reillz10 — APPROVED** [review]
Such a nice refactor! This makes everything feel way less janky.

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -1832,7 +1834,7 @@ class CurriculumCourseMembership(TimeStampedModel):
     Represents the Courses that compose the curriculum of a degree.
     """
     curriculum = models.ForeignKey(Curriculum, on_delete=models.CASCADE)
-    course = models.ForeignKey(Course, on_delete=models.CASCADE)
+    course = models.ForeignKey(Course, on_delete=models.CASCADE, related_name='program_curriculum_membership')
```
Comment (iloveagent57) [inline course_discovery/apps/course_metadata/models.py]: I think you want `curriculum_course_membership` as the related_name here.

File: course_discovery/apps/course_metadata/signals.py [hunk 4]
```diff
@@ -34,17 +39,53 @@ def add_masters_track_on_course(sender, instance, **kwargs):  # pylint: disable=
...
         for course_run in instance.course_runs:
-            _add_masters_track_on_commerce_course_run(course_run, partner.lms_commerce_api_url, commerce_api_client)
+            Seat.objects.update_or_create(
+                course_run=course_run,
+                type=Seat.MASTERS,
+                currency=us_currency,
+            )
+
+
+def get_linked_masters_program(course_run):
```
Comment (iloveagent57) [inline course_discovery/apps/course_metadata/signals.py]: I'm not going to leave this comment everywhere in this PR, but I don't like the word "linked" for what you mean here.  I'd much prefer we call them "related" masters programs.  "Linked" sounds like something very specific, or like it has to do with URL links.

File: course_discovery/apps/course_metadata/signals.py [hunk 5]
```diff
@@ -26,6 +27,10 @@ def delete_program(sender, instance, **kwargs):  # pylint: disable=unused-argume
         publisher.delete_obj(instance)
 
 
+def is_program_masters(program):
+    return program and program.type.slug == 'masters'
+
+
 @receiver(post_save, sender=CurriculumCourseMembership)
 def add_masters_track_on_course(sender, instance, **kwargs):  # pylint: disable=unused-argument
     # When this waffle flag is enabled we api call to ensure that the 'masters' course_mode is added to
```
Comment (iloveagent57) [inline course_discovery/apps/course_metadata/signals.py]: This comment is now out-of-date.

Comment (iloveagent57) [reply course_discovery/apps/course_metadata/signals.py]: Also, prefer a docstring.

File: course_discovery/apps/course_metadata/signals.py [hunk 6]
```diff
@@ -34,17 +39,53 @@ def add_masters_track_on_course(sender, instance, **kwargs):  # pylint: disable=
...
+def get_linked_masters_program(course_run):
+    curriculum_memberships = course_run.course.program_curriculum_membership.all()
+    for membership in curriculum_memberships:
+        if is_program_masters(membership.curriculum.program) and course_run in membership.course_runs:
+            return membership.curriculum.program
+
+
+@receiver(post_save, sender=Seat)
+def publish_masters_track(sender, instance, **kwargs):  # pylint: disable=unused-argument
+    # We should only publish the Masters track of the seat in the course_run,
```
Comment (iloveagent57) [inline course_discovery/apps/course_metadata/signals.py]: Can you turn this into a docstring and say a little more specifically what it does?  Maybe mention the commerce API?

File: course_discovery/apps/course_metadata/signals.py [hunk 7]
```diff
@@ -34,17 +39,53 @@ def add_masters_track_on_course(sender, instance, **kwargs):  # pylint: disable=
...
+    # if the course_run is linked to a Program of Masters type
+
+    if instance.type != Seat.MASTERS:
```
Comment (iloveagent57) [inline course_discovery/apps/course_metadata/signals.py]: some variable-renaming might make this cleaner.  If you do `seat = instance` above this, you can refer to `seat.type` and `seat.course_run` below.

**Review by iloveagent57 — COMMENTED** [review]
I have complaints about naming and style, all of the logic and testing looks good.

**Comment by schenedx** [comment]
@iloveagent57 I addressed all of your PR feedback. Can you review again?

**Review by iloveagent57 — APPROVED** [review]

**Pull request merged**